pyggi/base/program.py

```diff
@@ -0,0 +1,238 @@
+"""
+
+This module contains GranularityLevel and Program class.
+
+"""
+import os
+import shutil
+import json
+import time
+import pathlib
+import random
+import enum
+import collections
+import subprocess
+import shlex
+from copy import deepcopy
+from abc import ABC, abstractmethod
+from distutils.dir_util import copy_tree
+from . import InvalidPatchError
+from .. import PYGGI_DIR
+from ..utils import Logger
+
+class StatusCode(enum.Enum):
```

Comment:
I feel like there are at least three superposed mechanisms to check is a run was successful, in which case two are probably superfluous.
In the end, we only check is the fitness is not falsy (https://github.com/coinse/pyggi/blob/ref/pyggi/algorithms/local_search.py#L123), fitness being either None when the run was unsuccessful, or whatever "compute_fitness" returns otherwise.
Due to the number of times it bit me when I was playing in my branch, I feel that "not fitness" should be "fitness is None" for cases when for example the fitness is the number of failing test cases ("not 0" is True, but "0 is not None" is False).
If the first mechanism is to simply check if the fitness exists, it is superposed with the StatusCode enum which is ignored at the end (the test could use "status_code" instead).
At last, the two "not Normal" cases are time outs and parsing error, for which there already are exceptions (subprocess.TimeoutExpired which granted would need to be renamed, and InvalidPatchError).
InvalidPatchError is very interesting because is it only used once, to be immediately caught; it is called after the Patch class, defined in the Patch class, but only exist between two immediate consecutive functions of AbstractProgram.

Three possible solutions:
- have only "None" and "not None" fitnesses (does it really matter to know how the test failed? if yes I would rather simply return a dummy value, e.g., PAR10: 10 times the duration allowed running time).
- keep an enum based solution, but with fitness an attribute of the Normal member
- simply propagate the exceptions